src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,520 @@
+#include "deriveNativeScriptHash.h"
+#include "state.h"
+#include "uiScreens.h"
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+#define TRACE_WITH_CTX(message, ...) TRACE(message "level = %u, remaining scripts = %u", ##__VA_ARGS__, ctx->level, ctx->complexScripts[ctx->level].remainingScripts);
+#define SAFE_SPRINTF(str, format, ...) {\
+		explicit_bzero(str, SIZEOF(str));\
+		snprintf(str, SIZEOF(str), format, __VA_ARGS__);\
+		ASSERT(strlen(str) + 1 < SIZEOF(str));\
+	}
+
+static void expectsMoreScripts()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	VALIDATE(ctx->complexScripts[ctx->level].remainingScripts > 0, ERR_INVALID_STATE);
+}
+
+static inline void simpleScriptFinished()
+{
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("simple script finished, ");
+}
+
+static inline void complexScriptFinished()
+{
+	ASSERT(ctx->level > 0);
+	ctx->level--;
+
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("complex script finished, ");
+}
+
+// UI
+
+const char ui_native_script_header[7][27] = {"Script - key path", "Script - key", "Script - ALL", "Script - ANY", "Script - N of K", "Script - invalid before", "Script - invalid hereafter"};
+
+#define ASSERT_UI_SCRIPT_TYPE_SANITY() ASSERT(ctx->ui_scriptType >= UI_SCRIPT_PUBKEY_PATH && ctx->ui_scriptType <= UI_SCRIPT_INVALID_HEREAFTER);
+#define HEADER ui_native_script_header[ctx->ui_scriptType]
+
+#define SPRINTF_POSITION(offset, str) \
+	offset += snprintf(&str[offset], 10 + 1, "Position: ");\
+	for (int i = 1; i <= level; i++) {\
+		snprintf(&str[offset], 11 + 1, "%u.", ctx->complexScripts[i].totalScripts - ctx->complexScripts[i].remainingScripts + 1);\
+		/* snprintf returns 0, https://github.com/LedgerHQ/nanos-secure-sdk/issues/28 */\
+		offset += strlen(&str[offset]);\
+	}
+
+static void deriveScriptHash_display_ui_complexScriptFinished(ui_callback_fn_t* callback)
+{
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+	TRACE();
+
+	uint8_t level = ctx->level - 1;
+	if (level == 0) {
+		// Don't show position for complex script at level 0, "root" level
+		ui_displayPaginatedText(
+		        HEADER,
+		        "Finished",
+		        callback
+		);
+		return;
+	}
+
+	// "Position: x.x.x - Finished"
+	char content[10 + 11 * level + 10 + 1];
```

Comment:
I was thinking, because I saw that the struct that eventually contains the text for the UI, has the attribute `char fullText[200];`, which I guess means that eventually anything longer than 200 characters won't be shown.
However, theoretically such a string is possible here with some crazy nesting, it shouldn't happen, because the native scripts shouldn't be so complex, but how should I deal with this?
The length of this array should probably be also 200, but I should probably check for eventual overflows right?

Reply:
The overflows are, at present, caught by
```
	ASSERT(body_len < SIZEOF(ctx->fullText));
```
in `ui_displayPaginatedText`. Your code should not directly deal with the UI buffer length, but it would make sense if `uiHelpers.h` offered a function that would check header and/or second line lengths and throw ERR_DATA_TOO_LARGE if it did not fit. It is definitely better to propagate a particular error code than letting Ledger just freeze on some assertion.

Reply:
I tried to implement the solution you proposed :)